src/Chainweb/Mempool/Mempool.hs

```diff
@@ -10,8 +10,64 @@
 
 {-# OPTIONS_GHC -fno-warn-orphans #-}
 
+------------------------------------------------------------------------------
+-- | Mempool
+--
+-- The mempool contains an in-memory store of all of the pending transactions
+-- that have been submitted by users for inclusion into blocks, and is
+-- responsible for collecting the candidate transaction set when we prepare a
+-- new block for mining.
+--
+-- The in-memory mempool implementation (see @InMem.hs@ and @InMemTypes.hs@)
+-- contains three datasets:
+--
+-- 1. a priority search queue of pending transactions, keyed by pact
+-- transaction hash, and (currently) ordered by gas price (descending) with
+-- ties broken by gas limit (ascending). Basically -- we prefer highest bidder
+-- first, and all else being equal, smaller transactions.
+--
+-- 2. a \"quarantine\" map from hash to transaction, where we store
+-- transactions that were sent to us but haven't yet been validated. More on
+-- this below.
+--
+-- 3. a recent-modifications log, storing the last @k@ updates to the mempool.
+-- This is in here so peers can sync with us by fetching starting with a
+-- high-water mark, rather than having to go through a complete re-sync of all
+-- transaction hashes.
+--
+-- Transaction lifecycle:
+--
+--   - a transaction is created and signed by a user, and sent (via the pact
+--     @/send@ endpoint) to the mempool, using 'mempoolQuarantine'
+--
+--   - the transaction enters a /quarantine pool/ until it can be validated.
+--     Mempool will not gossip a transaction to peers until it has been
+--     validated.
```

Comment:
What exactly is the goal of "validation" here. 

Pact validation is "per block" is with respect to block in the chain (not a tx) and is expensive. A tx that is valid in one block might be invalid in another block. A tx has to be re-evaluated for each block into which it is included.

A tx is gossiped around before it is included in block. So it it's not clear what "validation" here means. If it refers to some wellformedness checks, let's call it that way, to avoid confusion with block validation.

Also, why is expensive pact validation needed for a wellformedness check? It introduced delays and circular dependencies with pact validation. Overall it seems to make the control flow related to mempool-pact interaction much more complex.

Reply:
A propose that (as a followup to this PR) we assemble a list of wellformedness checks that a tx should pass in order to be included in the pool of pending/gossiped transactions.

Once we have that list it will be clear what resources are needed to perform those checks, in particular, whether pact-service is needed at all for that.

Reply:
You are correct that this is basically a well-formedness check, and is intended to be quick. I can rename it. The well-formedness criteria as they're sketched out right now:

  - the transaction has not already been played in a block (this requires the precise transaction index)
  - the transaction is not expired relative to the block's clock time (not yet implemented)
  - the transaction has a valid signature for a coin sender (Emily is still working on this)
  - the sender's account has at least the gas limit amount of coin (we think we have this working)

You are correct that we will drop transactions if (for example) consensus briefly lands on a fork where you don't have enough gas. Tough! This means that `sum(gasLimit)` for all your outstanding transactions must exceed your account balance. The reasons we are piggybacking the initial mempool well-formedness check onto newBlock time: a) performance (so that we avoid moving the pact db pointer, which is expensive), b) this gives us a convenient synchronization point with consensus.

Reply:
OK, that makes sense.  Some of the checks seem relatively strong for wellformedness:

> the transaction has not already been played in a block (this requires the precise transaction index)
> the sender's account has at least the gas limit amount of coin (we think we have this working)

I would expect that wellformedness does't depend on a context, but these two checks depend on the context of a particular block.

> the transaction is not expired relative to the block's clock time

This also depends on the context of a block. For wellformedness, wouldn't it be enough to check the weaker conditions `TTL < maxTTL` and `now <= creationTime tx + TLL <= now`?

> the transaction has a valid signature for a coin sender

What does this include? Just validity of the signature also authentication with respect to some data/account in the chain?

I would expect that wellformedness also includes more basic checks like:

* the tx has can be parsed and contains syntactically valid pact.
* the tx has the correct chainweb version.
* the tx is for the correct chain.
* the tx has a supported pact version.
* the tx fits into, yet to be defined, broad syntactic limits (e.g. overall size in bytes).
* the tx has acceptable (with respect to some broad constants) gas limits.
* the sender (public key) of the tx isn't blacklisted.

I assume some of these are probably implicit by successfully decoding the tx. It may still be good to document this.

Regarding piggybacking onto pact new block validation, I am not sure about the pros and cons. It seems to me that piggy packing adds some complexity to the code, and also causes tx to fail relatively late, after already causing some resource consumption. It also seems problematic that the gossip behavior depends on the checkpoint and the miner calling newBlock. But I also see how gossiping only after more rigorous validation helps to prevent DOS attacks.

There might be some middle ground here with (1) a cheap fail-fast wellformedness check for incoming tx and (2) an API for feedback from pact validation about tx validation failures that could be used for P2P reputation management for throwing and/or blacklisting senders of malformed txs or txs that consistently fail pact validation.

Reply:
I'll rip out the quarantine later today since Stu doesn't want it. Will add a TODO to add the pre-gossip check